### ISSUE TITLE
[Enhancement] Add support for elements containers lists, like: List<ElementContainer> = $$("selector")

### ISSUE THREAD

**Issue by savvagen** [selenide/selenide#1225]
It would be very helpfull to build "liststs of element containers" without building custom wrappers.
Example:  
Building element container list

class ProductCard extends <ElementsContainerCollection> {
    SelenideElement title = $("title")
    SelenideElement price = $("price")
} 

then

class CatalogPage{
   List<ProductCard> products = $$("ul#catalog li") 
}

**Comment by SeleniumTestAB** [comment]
Question! Does java even allow such thing? I know JS and C# (to some extend) allow such syntax, but i am not sure if java does (please correct me if wrong)

**Comment by savvagen** [comment]
> Question! Does java even allow such thing? I know JS and C# (to some extend) allow such syntax, but i am not sure if java does (please correct me if wrong)

Yes, such thing can be ealized by creating your own Element Containers list, like in such example, which can be used in Elements pattern:

1. Element Container
```
class ProductCard {

    SelenideElement el;
   
    public ProductCard(ElementsCollection el){
        this.el = el;
    }
    
    SelenideElement title = $("#title")
    SelenideElement price = $("#price")
} 
```
2. Container List

```
class ProductList {

    ElementsCollection el;

    public ProductList(ElementsCollection el){
        this.el = el;
    }
    
    public ProductCard get(int index){
        return new ProductCard(el.get(index))
    }
    
    public ProductList shouldHaveSize(int size){
       el.shouldHaveSize(size)
       return this
    }
    
    
} 
```
3. Create list in the page objects:

class CatalogPage{
   ProductList products = new ProductList($$("ul#catalog li"))

  @Test 
  public void shouldTestList(){
       catalogPage.products.get(0).title.shouldHave(text("First Ttitle"))
  }
}

**Comment by SeleniumTestAB** [comment]
Yes, i see... i was wondering about that part:
`List<ProductCard> products = $$("ul#catalog li")`
because thats not a java syntax (kinda). but if you say it like:
`List<ProductCard> products = new ProductCards($$("ul#catalog li")).get()`

than yea that is somewhat nice.

Also i do see you got ElementsContainer, aint that good enough?

**Comment by savvagen** [comment]
> Yes, i see... i was wondering about that part:
> `List<ProductCard> products = $$("ul#catalog li")`
> because thats not a java syntax (kinda). but if you say it like:
> `List<ProductCard> products = new ProductCards($$("ul#catalog li")).get()`
> 
> than yea that is somewhat nice.
> 
> Also i do see you got ElementsContainer, aint that good enough?
> 
> P.S. now i did realize that you made those ElementContainer class by yourself. So the feature might be good (but i did see such things in selenide code aswell...)

Yea) In Selenide we have "ElementContainer" that works with PageFactory.
Now it will nice to have the same ElementContainer feature, that works without PageFactory, because while started working with Selenide, I am not using PageFactory anymore :)

**Comment by asolntsev** [comment]
Hi @savvagen !
Do I correctly understand that in your example, elements `#title` and `#price` should be searched inside of `ul#catalog li` (each of them)? 

I am afraid that it's impossible to implement your request before https://github.com/selenide/selenide/issues/1048 is done - but it seems hard or even impossible. :(

**Comment by savvagen** [comment]
> Hi @savvagen !
> Do I correctly understand that in your example, elements `#title` and `#price` should be searched inside of `ul#catalog li` (each of them)?
> 
> I am afraid that it's impossible to implement your request before #1048 is done - but it seems hard or even impossible. :(

Hello! Due to you question and example before, - the answer is "yes". It requires ability to find elements collection `ul#catalog li` and then search `#title` and `#price` in each item of the collection 

If it is possible, I think the solution with generics also may be nice, but hard for implementation:

```
        ...
        ElementsCollection<ProductCard> products = Selenide.$$("ul#catalog li");
        ...
        
        class ProductCard extends <SelenideElementContainer> {}

```

**Comment by Omer3160** [comment]
Hi, 
Is this issue going to be closed soon or is there an alternative solution?

**Comment by asolntsev** [comment]
@Omer3160 If somebody wishes to implement this feature - they are welcome. 
But I personally rather think this feature is not really needed. It's easy to create page objects, element containers etc. with just plain Java, without any special Selenide features.

**Comment by kfirdo** [comment]
@Omer3160 @savvagen Might be late to the party but the following code seems to do the basics:
VideoCard.java
```Java
import com.codeborne.selenide.ElementsContainer;
import com.codeborne.selenide.SelenideElement;

public class VideoCard extends ElementsContainer {
    private SelenideElement context;

    public VideoCard(SelenideElement element) {
        this.context = element;
    }

    public String getLink() {
        return context.find("#thumbnail").getAttribute("href");
    }
}
```

Application.java
```Java
import com.codeborne.selenide.ElementsCollection;
import com.codeborne.selenide.ElementsContainer;
import com.codeborne.selenide.Selenide;
import com.codeborne.selenide.SelenideElement;
import lombok.SneakyThrows;
import org.openqa.selenium.By;

import java.util.List;
import java.util.stream.Collectors;

public class Application {

    public static void main(String[] args) {
        Selenide.open("https://www.youtube.com/");

        List<VideoCard> cards = $$(By.cssSelector("ytd-rich-item-renderer"), VideoCard.class);

        for (VideoCard card : cards) {
            System.out.println(card.getLink());
        }
    }

    public static <T extends ElementsContainer> List<T> $$(By selector, Class<T> type) {
        ElementsCollection elements = Selenide.$$(selector);

        return elements.stream()
                       .map(element -> mapElement(element, type))
                       .collect(Collectors.toList());
    }

    @SneakyThrows
    private static <T extends ElementsContainer> T mapElement(SelenideElement element, Class<T> customType) {
        return customType.getConstructor(SelenideElement.class).newInstance(element);
    }
}

```
This will provide a more 'generic' solution to @savvagen, more logic can be added to even create custom lists such as 'ProductList' but I wanted to keep it simple.
Does that achieve what you wanted?

**Comment by asolntsev** [comment]
Thank you @kfirdo!
That's what I am talking about.

Actually, VideoCard doesn't need to extend ElementsContainer. Just remove "
extends ElementsContainer. :)

As you see, it works without any special support from selenide side. Just
plain old good _programming_. :)

пн, 16 авг. 2021 г., 17:15 kfirdo ***@***.***>:
